Better processing of example data prior to creating dataset component

freddyaboulton opened the pull request:
# Description
The goal of this PR is to add some processing to the example data passed in by users to either:

1. Accept more data formats as valid examples for a given component, e.g. pandas dataframes for dataframes
2. Improve how the data is displayed in the UI, e.g. only display the filename for a file as opposed to the full path
3. Validate the data matches the expected format by the UI, e.g. an image example must be a path to an image file.

Points 1 and 2 follow my discussion with @abidlabs [here](https://huggingface.slack.com/archives/C02SPHC1KD1/p1661808604873899)

Point 3 is **not implemented** since I did not discuss that with anyone prior  but I think will improve the developer experience. Unfortunately, I don't think we'll be able to validate file-path based components due to the ambiguity between absolute paths and relative paths when reading from a csv file: #2129

In terms of design, we opted to add an `as_example` method that will be called by the Dataset component on init. Didn't want to couple the component-specific example logic with the dataset component implementation in case the developer is defining their own component. We might have to re-work how we do examples when we get to community components though.

You can test the original dataframe issue here:

```python
import gradio as gr
import pandas as pd


examples = [[pd.DataFrame([[7, 8, 9], [10, 11, 12]])],
             [pd.DataFrame([[5, 6, 7], [8, 9, 10]])]]


def get_max(df: pd.DataFrame, *args) -> float:
    return df.max().max()


gr.Interface(get_max, [gr.DataFrame(type="pandas"), gr.Number()], gr.Number(),
             examples=examples).launch()
```

Closes: #2089



# Checklist:

- [x] I have performed a self-review of my own code
- [x] My code follows the style guidelines of this project
- [x] I have commented my code in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

freddyaboulton commented on gradio/components.py:
```diff
@@ -210,6 +211,10 @@ def get_load_fn_and_initial_value(value):
         return load_fn, initial_value
 
+    @abstractmethod
+    def as_example(self, input_data):
```
Marking this as an abstract method so we don't forget to add it for new components. The trade-off is that there will be some repeated 

```python
    def as_example(self, input_data):
        return input_data
```

code but I think it's worth it for the "safety" of knowing a someone had to type out an implementation as opposed to just forgetting and inheriting from IO component.

freddyaboulton commented on gradio/components.py:
```diff
@@ -1506,6 +1532,15 @@ def stream(
             _postprocess=_postprocess,
         )
 
+    def as_example(self, input_data):
```
We can add this validation to other components to improve the developer experience, e.g. Audio, Video, CheckboxGroup. We can also get really detailed if we want, e.g. Image files must end in png or jpg

freddyaboulton commented:
@abidlabs Tagging you for review even though it's still draft just cause I want to make sure this is on the right path

abidlabs commented:
This looks great to me @freddyaboulton! I actually think it's fine to make `as_example()` a regular method (not an abstract method) defined in the `IOComponent` class. The default implementation should be the identity function.

Yes, we have to remember to implement it for all of the `IOComponent` subclasses, but the same is true for `preprocess()`, `postprocess()`, or any data-processing method. And given that most components' `as_example()` method is simply the identity function, this will save a lot of repeated code.

abidlabs commented:
I'll look into the filepath problem and the issue you created

freddyaboulton commented on gradio/examples.py:
```diff
@@ -6,10 +6,9 @@
 import csv
 import inspect
 import os
-import shutil
```
Noticed not used

freddyaboulton commented:
@abidlabs This should be good for review now. Removed the abstract method and added tests. We can customize the `as_example` for different components in the future, e.g. accepting PIL images as mentioned [here](https://huggingface.slack.com/archives/C02QZLG8GMN/p1661930679607899)

abidlabs commented:
LGTM @freddyaboulton, tested and works great!
